### ISSUE TITLE
你好，我用 pytorch 版本的 XLNet-base进行测试生成，未 fine-tuning，发现效果贼差，不知道怎么回事？

### ISSUE THREAD

**Issue by wulaoshi** [ymcui/Chinese-XLNet#53]
![Uploading image.png…]()

**Comment by wulaoshi** [comment]
关键代码，大致这样：
`tokenizer = XLNetTokenizer.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/spiece.model")
model = Model.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/")
current_output = []
utt_history = ["你好，你叫什么名字？"]
    for i in range(MAX_INPUT_LENGTH-10):
        utt_history = [tokenizer.convert_tokens_to_ids(tokenizer.tokenize(ht)) for ht in utt_history + current_output]
        input_ids = torch.tensor(utt_history, device=self.device).unsqueeze(0)
        logits = self.generator_model(input_ids)
        if isinstance(logits, tuple):
            logits = logits[0]
        logits = top_filtering(logits, top_k=top_k, top_p=top_p)
        probs = F.softmax(logits, dim=-1)
        prev = torch.topk(probs, 1)[1]
        if i < self.min_length and prev.item() in special_tokens_ids:
            while prev.item() in special_tokens_ids:
                if probs.max().item() == 1:
                    print("Warning: model generating special token with probability 1.")
                    break  #避免一直生成
                prev = torch.multinomial(probs, num_samples=1)
        if prev.item() in special_tokens_ids:
            break
        current_output.append(prev.item())
    out_text = self.generator_tokenizer.decode(current_output, skip_special_tokens=True)`

**Comment by wulaoshi** [comment]
tokenizer = XLNetTokenizer.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/spiece.model")
model = Model.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/")
current_output = []
utt_history = ["你好，你叫什么名字？"]
    for i in range(MAX_INPUT_LENGTH-10):
        utt_history = [tokenizer.convert_tokens_to_ids(tokenizer.tokenize(ht)) for ht in utt_history + current_output]
        input_ids = torch.tensor(utt_history, device=self.device).unsqueeze(0)
        logits = self.generator_model(input_ids)
        if isinstance(logits, tuple):
            logits = logits[0]
        logits = top_filtering(logits, top_k=top_k, top_p=top_p)
        probs = F.softmax(logits, dim=-1)
        prev = torch.topk(probs, 1)[1]
        if i < self.min_length and prev.item() in special_tokens_ids:
            while prev.item() in special_tokens_ids:
                if probs.max().item() == 1:
                    print("Warning: model generating special token with probability 1.")
                    break  #避免一直生成
                prev = torch.multinomial(probs, num_samples=1)
        if prev.item() in special_tokens_ids:
            break
        current_output.append(prev.item())
    out_text = self.generator_tokenizer.decode(current_output, skip_special_tokens=True)

**Comment by wulaoshi** [comment]
```
tokenizer = XLNetTokenizer.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/spiece.model")
model = Model.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/")
current_output = []
utt_history = ["你好，你叫什么名字？"]
    for i in range(MAX_INPUT_LENGTH-10):
        utt_history = [tokenizer.convert_tokens_to_ids(tokenizer.tokenize(ht)) for ht in utt_history + current_output]
        input_ids = torch.tensor(utt_history, device=self.device).unsqueeze(0)
        logits = self.generator_model(input_ids)
        if isinstance(logits, tuple):
            logits = logits[0]
        logits = top_filtering(logits, top_k=top_k, top_p=top_p)
        probs = F.softmax(logits, dim=-1)
        prev = torch.topk(probs, 1)[1]
        if i < self.min_length and prev.item() in special_tokens_ids:
            while prev.item() in special_tokens_ids:
                if probs.max().item() == 1:
                    print("Warning: model generating special token with probability 1.")
                    break  #避免一直生成
                prev = torch.multinomial(probs, num_samples=1)
        if prev.item() in special_tokens_ids:
            break
        current_output.append(prev.item())
    out_text = self.generator_tokenizer.decode(current_output, skip_special_tokens=True)
```

**Comment by wulaoshi** [comment]
难道 xlnet 不能用来做生成任务吗？

**Comment by ymcui** [comment]
1. 一楼的图片看不到
2. XLNet做生成，我这边没有试过，你可以用官方的英文权重试一下，看看是否有同样的问题。
3. XLNet官方目录中做文本生成的相关讨论，供参考：https://github.com/zihangdai/xlnet/issues/135 ， https://github.com/zihangdai/xlnet/issues/94

**Comment by wulaoshi** [comment]
![123](https://user-images.githubusercontent.com/27938964/89020497-98f18c00-d351-11ea-917e-995e39fb6d5e.png)
你好，这是生成的回复

**Comment by wulaoshi** [comment]
这是测试代码
```
tokenizer = XLNetTokenizer.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/spiece.model")
model = Model.from_pretrained(pretrained_models_path+"chinese_xlnet_base_pytorch/")
current_output = []
utt_history = ["你好，你叫什么名字？"]
    for i in range(MAX_INPUT_LENGTH-10):
        utt_history = [tokenizer.convert_tokens_to_ids(tokenizer.tokenize(ht)) for ht in utt_history + current_output]
        input_ids = torch.tensor([utt_history[0] + current_output], device=self.device)
        logits = self.generator_model(input_ids)
        if isinstance(logits, tuple):
            logits = logits[0]
        logits = top_filtering(logits, top_k=top_k, top_p=top_p)
        probs = F.softmax(logits, dim=-1)
        prev = torch.topk(probs, 1)[1]
        if i < self.min_length and prev.item() in special_tokens_ids:
            while prev.item() in special_tokens_ids:
                if probs.max().item() == 1:
                    print("Warning: model generating special token with probability 1.")
                    break  #避免一直生成
                prev = torch.multinomial(probs, num_samples=1)
        if prev.item() in special_tokens_ids:
            break
        current_output.append(prev.item())
    out_text = self.generator_tokenizer.decode(current_output, skip_special_tokens=True)
```

**Comment by ymcui** [comment]
不好意思，生成方面我确实没法提供一些建议。
还是建议你去XLNet官方目录以'generation'为关键词查找其他人的一些实践方案。
另外，还是建议你先在英文的预训练模型上进行尝试，以排除是否是中文模型本身的问题导致效果较差。

**Comment by wulaoshi** [comment]
好的，谢谢
